### ISSUE TITLE
retry failed

### ISSUE THREAD

**Issue by rotten** [komamitsu/treasuredata_fdw#41]
I have a materialized view on top of a foreign data wrapper.

When I was using td_fdw, I would refresh the materialized view, it would get to 74M rows (out of the 114M rows) and complete without any warning or error.  Missing 35% of our data was not going to cut it.

I was able to reduce the table size to 94M rows, and switch to the treasuredata_fdw.  Now when I create the materialized view it runs for 12 hours or so, and then throws this error:

`ERROR:  issue_query: Failed to fetch result. job_id=174387897, error=RetryError { message: "reached last try without condition match" }`

At that point the create materialized view fails.  I can't tell if it is failing at the same 74M row mark or not because it is a create instead of a refresh, and thus rolls back then entire table, but it wouldn't surprise me.

Have I reached some sort of maximum query size?  I am not sure if the problem is in the FDW or postgresql yet.  I am not running out of temp space, table space, or memory in the database server as near as I can tell.  I can work on partitioning the materialized view into smaller pieces if necessary, but I'd like to make sure that is what is required to get this data into postgresql and that there isn't some other issue at play here.

**Comment by rotten** [comment]
I tried it again last night, just in case it was a random glitch.  Nope, same error:

**Comment by rotten** [comment]
I'm using the 'Hive' query engine.  I'll try a Presto run to see what happens.  It takes about 12 hours before I get the failure.

**Comment by komamitsu** [comment]
Thanks for the reporting.
The failure occurred in the fdw not in Postgres.

I don't think there is an explicit query size limitation. But actually it can fail due to network issue or etc.

BTW, I noticed there was room to improve the retry functions of fetching result data from Treasure Data in the fdw and I addressed it in version 1.2.9. Could you try 1.2.9 when you get a chance? If it fails again, an error message with the job id would help for investigation.

**Comment by rotten** [comment]
Will do!  I bumped the treasuredata_fdw extension on the server.  Unfortunately it is hard to tell if the database picks it up when I do:
```
clarivoy=# alter extension treasuredata_fdw update;
NOTICE:  version "1.2" of extension "treasuredata_fdw" is already installed
ALTER EXTENSION
```
Last time I was able to simply drop and recreate the extension, that won't work this time because of other dependencies.

When I do a "\dx" it only shows version 1.2  (unlike other packages such as postgis which show 2.3.3 with the extra point release in the version string).

This doesn't help either:
```
=# select * from pg_available_extensions where name = 'treasuredata_fdw';
       name       | default_version | installed_version |                comment                 
------------------+-----------------+-------------------+----------------------------------------
 treasuredata_fdw | 1.2             | 1.2               | foreign-data wrapper for Treasure Data
(1 row)

```
I'm going to assume it was safely bumped from 1.2.8 to 1.2.9 and try the run again (with Hive).  I'll let  you know how it goes.

**Comment by rotten** [comment]
Well - that died a lot sooner than 12 hours in:

`ERROR:  Failed to fetch result. job_id=175078279, error=IoError(Error { repr: Custom(Custom { kind: Other, error: StringError("early eof") }) })`

**Comment by komamitsu** [comment]
I see. Thank you for trying it.

It looks like something related to network issue. I'll try to reproduce it.

BTW,

> Well - that died a lot sooner than 12 hours in:

how long did it take to fail?

**Comment by rotten** [comment]
It took about an hour.

**Comment by komamitsu** [comment]
I saw `td-client-rust` itself failed to fetch large query results from Treasure Data when the size was more than about 6.5GB. After some investigations of server side implementations, I found there are 2 modes on the server side to handle a fetch request from td-client and `td-client-rust` should use the other one.

I created a new branch `improve-result-fetch` that uses the better mode for downloading a query result and I hope it resolves this issue. Could you try the branch when you get a chance?

**Comment by rotten** [comment]
I installed the improve-result-fetch version of the FDW.  I cannot bounce my production database at this time, and as usual, I couldn't tell if alter extension update picked up the patch because the version number didn't change.

After running the create materialized view again, it died after an hour with the same error:

`ERROR:  Failed to fetch result. job_id=175598426, error=MsgpackDecodeError(InvalidDataRead(Error { repr: Custom(Custom { kind: Other, error: StringError("early eof") }) }))`

**Comment by komamitsu** [comment]
Hmm... I can't reproduce this issue with the branch on our dataset and instances...

I think `improve-result-fetch` fixes a problem in downloading a result, but I guess there is another problem related to your network or something.

I just created a github project https://github.com/komamitsu/tdclient-dltest. As you may be able to see in the source code, it just downloads a query result with a proper wait.

I want you to try the application as follows and let's see if it keeps running for a several hours not only one hour to isolate the cause.

```
$ git clone https://github.com/komamitsu/tdclient-dltest.git
$ cd tdclient-dltest
$ cargo build --release
$ cargo run --release $(your_apikey) $(job_id. In this case, let's try 175078279)
# For instance, cargo run --release "99999/sdfaskjfhaksjdfhla" 175078279
```

**Comment by rotten** [comment]
```
 cargo run --release 'xxx' 175078279
    Finished release [optimized] target(s) in 0.0 secs
     Running `target/release/tdclient-dltest xxx 175078279`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: MsgpackDecodeError(InvalidDataRead(Error { repr: Custom(Custom { kind: Other, error: StringError("early eof") }) }))', /checkout/src/libcore/result.rs:860:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.

```

**Comment by rotten** [comment]
It took 15 minutes or so for it to throw that error.  I'm not sure what could be wrong with the network, our servers are running in the Google Cloud and have never shown any other network issues.

**Comment by komamitsu** [comment]
The master branch of https://github.com/komamitsu/tdclient-dltest uses:
- Old `td-client-rust` that doesn't contain the change to improve fetching query results
- A result fetch API with deserializing MessagePack format not only downloading

I just created additional 3 branches to isolate the problem (`without-deser`, `improve-retry-fetch`, `without-deser-and-improve-retry-fetch`). Could you try these branches like this when you have a chance?

```
$ pwd
path/foo/bar/tdclient-dltest
$ git fetch origin
$ git checkout without-deser
$ cargo build --release
$ cargo run --release $(your_apikey) $(job_id. In this case, let's try 175078279)
# For instance, cargo run --release "99999/sdfaskjfhaksjdfhla" 175078279

# And then, let's try `improve-retry-fetch` and `without-deser-and-improve-retry-fetch`
```

**Comment by rotten** [comment]
First try "without-deser":

```
$  git checkout without-deser
Branch without-deser set up to track remote branch without-deser from origin.
Switched to a new branch 'without-deser'
$   cargo build --release
   Compiling tdclient-dltest v0.1.0 (file:///home/rotten/src/tdclient-dltest)
    Finished release [optimized] target(s) in 6.52 secs
$  cargo run --release 'xxx' 175078279
    Finished release [optimized] target(s) in 0.0 secs
     Running `target/release/tdclient-dltest xxx 175078279`
finish
```
It took about 15-20 minutes.  No errors.  I'll try the next one, "improve-retry-fetch" and report back shortly.

**Comment by rotten** [comment]
This time with "imrpove-retry-fetch":

```
$  git checkout improve-retry-fetch
Branch improve-retry-fetch set up to track remote branch improve-retry-fetch from origin.
Switched to a new branch 'improve-retry-fetch'

$  cargo build --release
   Compiling td-client v0.6.0 (https://github.com/komamitsu/td-client-rust.git?branch=accept-encoding#a3335cc2)
   Compiling tdclient-dltest v0.1.0 (file:///home/rotten/src/tdclient-dltest)
    Finished release [optimized] target(s) in 9.20 secs

$  cargo run --release 'xxx' 175078279
    Finished release [optimized] target(s) in 0.0 secs
     Running `target/release/tdclient-dltest xxx 175078279`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: MsgpackDecodeError(InvalidDataRead(Error { repr: Custom(Custom { kind: Other, error: StringError("early eof") }) }))', /checkout/src/libcore/result.rs:860:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by rotten** [comment]
And the third one:

```
$  git checkout without-deser-and-improve-retry-fetch
Branch without-deser-and-improve-retry-fetch set up to track remote branch without-deser-and-improve-retry-fetch from origin.
Switched to a new branch 'without-deser-and-improve-retry-fetch'

$  cargo build --release
   Compiling tdclient-dltest v0.1.0 (file:///home/rotten/src/tdclient-dltest)
    Finished release [optimized] target(s) in 2.39 secs

$  cargo run --release 'xxx' 175078279
    Finished release [optimized] target(s) in 0.0 secs
     Running `target/release/tdclient-dltest xxx 175078279`
finish
```
This took about 10 minutes.

**Comment by komamitsu** [comment]
Thank you.

According to the results, we learned
1. Deserialization of MessagePack formatted data can be the cause.
2. `improve-retry-fetch` doesn't resolve this issue, but it improves the performance.

I'll look into the deserialization of MessagePack later.

**Comment by komamitsu** [comment]
I reproduced the error with `tdclient-dltest` capturing TCP sequences.

```
13:09:30.071709 IP ec2-server-host.compute-1.amazonaws.com.https > ip-client-host.us-west-2.compute.internal.48056: Flags [P.], seq 544217746:544238018, ack 3185196846, win 114, options [nop,nop,TS val 9446063 ecr 3419604857], length 20272
13:09:30.606545 IP ec2-server-host.compute-1.amazonaws.com.https > ip-client-host.us-west-2.compute.internal.48056: Flags [P.], seq 2849752:2887312, ack 1, win 114, options [nop,nop,TS val 9446199 ecr 3419604993], length 37560
    :
13:34:18.431369 IP ec2-server-host.compute-1.amazonaws.com.https > ip-client-host.us-west-2.compute.internal.48056: Flags [P.], seq 407051232:407052680, ack 1, win 114, options [nop,nop,TS val 9818155 ecr 3419976950], length 1448
13:34:18.912050 IP ec2-server-host.compute-1.amazonaws.com.https > ip-client-host.us-west-2.compute.internal.48056: Flags [P.], seq 409809672:409854560, ack 1, win 114, options [nop,nop,TS val 9818275 ecr 3419977070], length 44888
13:34:19.347971 IP ec2-server-host.compute-1.amazonaws.com.https > ip-client-host.us-west-2.compute.internal.48056: Flags [P.], seq 412785832:412786760, ack 1, win 114, options [nop,nop,TS val 9818385 ecr 3419977180], length 928
13:34:19.745021 IP ec2-server-host.compute-1.amazonaws.com.https > ip-client-host.us-west-2.compute.internal.48056: Flags [FP.], seq 415283112:415284378, ack 1, win 114, options [nop,nop,TS val 9818484 ecr 3419977279], length 1266
13:34:21.589505 IP ip-client-host.us-west-2.compute.internal.48056 > ec2-server-host.compute-1.amazonaws.com.https: Flags [F.], seq 1, ack 415284379, win
```

According the result, it seems the server side including AWS network stack disconnected the connection...

**Comment by rotten** [comment]
So some sort of firewall connection timeout.

**Comment by komamitsu** [comment]
- The Hive query via treasuredata_fdw after improving retry, which removes meaningless retries, ran for about 1 hour and failed.
- Just downloading the query result with deserialization ran for around 40 min and failed.

If the Hive query itself on TD took around 20 min, the result download took 40 min and failed. So we might be able to say "A connection running for about 40 min is disconnected by server side".

Oh, in the case of https://github.com/komamitsu/treasuredata_fdw/issues/41#issuecomment-330184653, it took only 15 min not 40 min...

**Comment by komamitsu** [comment]
Anyway, I have a plan to add a new attribute `query_download_dir` to treasuredata_fdw foreign table. If it's specified, the FDW downloads a query result file to local disk first and then deserializes result rows. With this feature, the duration of download a result file would be reduced and it could avoid this kind of problems although it consumes disk space temporally.

**Comment by rotten** [comment]
Sounds good to me!  I've temporarily partitioned the data into smaller chunks so I could get it loaded and meet the immediate business needs.  Let me know when this new option is ready and I'll give it a try and undo the temporary partitions.

**Comment by komamitsu** [comment]
@rotten I implemented a new option `query_download_dir` on https://github.com/komamitsu/treasuredata_fdw/tree/query-download-dir

Could you try the branch `query-download-dir` when you get a chance?

**Comment by rotten** [comment]
I'll give it a try in a few minutes. :-)  thanks!!

**Comment by rotten** [comment]
It worked great!  It dropped a file, and then loaded it into the database.  It took about 3 1/2 hours to complete successfully.  :-)   (with the old multicorn approach, recall, it was taking 24-30 hours before quietly finishing without indication something was wrong -- and when we started this process it was taking 12 hours before it failed with the rust fdw)   :+1: :100: :1st_place_medal:

**Comment by komamitsu** [comment]
Glad to hear that 🎉 

I'll release a new version that contains these changes later.

**Comment by komamitsu** [comment]
@rotten Released 1.2.11 that contains the fix.

**Comment by komamitsu** [comment]
I'm closing this issue. If you still have any problem, feel free to reopen :)
